### PR TITLE
[Merged by Bors] - test(systest): improve systest resilience

### PR DESCRIPTION
**Pull request by acud**
## Motivation

Improve systest resilience by retrying some of the checks on the `Unavailable` error.

## TODO

- [ ] Explain motivation or link existing issue(s)
- [ ] Test changes and document test plan
- [ ] Update documentation as needed
- [ ] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

File: systest/testcontext/context.go [hunk 1]
```diff
@@ -287,8 +291,16 @@ func New(t *testing.T, opts ...Opt) *Context {
...
 	}
+
+	t.Cleanup(func() {
+		if t.Failed() {
+			failOnce.Do(func() { close(failed) })
+		}
+	})
 	config, err := rest.InClusterConfig()
 	require.NoError(t, err)
+	config.Burst = 20
```
Comment (pigmej) [inline systest/testcontext/context.go]: this should be already set in https://github.com/spacemeshos/go-spacemesh/pull/6178/files#diff-ab921146878b38148b9a7082a00bd8c8645357f6fbf6b8e233fe83c93eff350dR294 if I'm not mistaken.

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Comment by pigmej** [comment]
bors try

**Comment by pigmej** [comment]
bors try

**Comment by pigmej** [comment]
bors try

**Comment by pigmej** [comment]
bors try

**Comment by acud** [comment]
bors try

File: systest/Makefile [hunk 2]
```diff
@@ -36,13 +36,15 @@ ifeq ($(configname),$(test_job_name))
...
 endif
 
-command := tests -test.v -test.count=$(count) -test.timeout=0 -test.run=$(test_name) -clusters=$(clusters) \
--level=$(level) -configname=$(configname)
+command := tests -test.v -test.count=$(count) -test.timeout=60m -test.run=$(test_name) -test.parallel=$(clusters) \
+					 -clusters=$(clusters) -level=$(level) -configname=$(configname)
 
 ifeq ($(failfast),true)
 	command := $(command) -test.failfast
 endif
```
Comment (fasmat) [inline systest/Makefile]: NIT: we don't need the if I believe and can just pass the value to the CLI

```suggestion
command := tests -test.v -test.count=$(count) -test.timeout=60m -test.run=$(test_name) -test.parallel=$(clusters) \
	-test.failfast=$(failfast) -clusters=$(clusters) -level=$(level) -configname=$(configname)
```

File: systest/cluster/cluster.go [hunk 3]
```diff
@@ -150,6 +150,9 @@ func ReuseWait(cctx *testcontext.Context, opts ...Opt) (*Cluster, error) {
 	if err := cl.WaitAllTimeout(cctx.BootstrapDuration); err != nil {
 		return nil, err
 	}
+	if err = cctx.CheckFail(); err != nil {
+		return nil, err
+	}
```
Comment (fasmat) [inline systest/cluster/cluster.go:155]: I don't think this is necessary? `failfast` already handles that no new tests will be started when the first one failed. Already running tests will continue to completion even with this check.

File: systest/testcontext/context.go [hunk 4]
```diff
@@ -2,6 +2,7 @@ package testcontext
 
 import (
 	"context"
+	goerr "errors"
```
Comment (fasmat) [inline systest/testcontext/context.go]: I prefer importing the std packages with their original names and renaming external packages if there are name collisions. Wdyt?

File: systest/tests/common.go [hunk 5]
```diff
@@ -196,7 +195,12 @@ func layersStream(
 		if ok && s.Code() != codes.OK {
 			logger.Warn("layers stream error", zap.String("client", node.Name), zap.Error(err), zap.Any("status", s))
 			if s.Code() == codes.Unavailable {
-				return nil
+				if retries == attempts {
+					return errors.New("layer stream unavailable")
+				}
+				retries++
+				time.Sleep(retryBackoff)
+				goto BACKOFF
```
Comment (fasmat) [inline systest/tests/common.go:203]: I think we should avoid implement this ourselves. `grpc.NewClient` can be used with `grpc.WithBackoffConfig()`. You could add that to `ensurePubCon` and `ensurePrivCon` in `node.go` and not need to use this loop to retry automatically. 🙂

File: systest/tests/common.go [hunk 6]
```diff
@@ -229,7 +236,13 @@ func malfeasanceStream(
...
 			)
 			if s.Code() == codes.Unavailable {
-				return nil
+				if retries == attempts {
+					return errors.New("layer stream unavailable")
+				}
+				retries++
+				time.Sleep(retryBackoff)
+				goto BACKOFF
+
```
Comment (fasmat) [inline systest/tests/common.go:245]: same as above

File: systest/tests/common.go [hunk 7]
```diff
@@ -324,7 +340,12 @@ func watchTransactionResults(ctx context.Context,
 					zap.Any("status", s),
 				)
 				if s.Code() == codes.Unavailable {
-					return nil
+					if retries == attempts {
+						return errors.New("transaction results unavailable")
+					}
+					retries++
+					time.Sleep(retryBackoff)
+					goto BACKOFF
```
Comment (fasmat) [inline systest/tests/common.go:348]: same as above

File: systest/tests/common.go [hunk 8]
```diff
@@ -360,7 +383,12 @@ func watchProposals(
 					zap.Any("status", s),
 				)
 				if s.Code() == codes.Unavailable {
-					return nil
+					if retries == attempts {
+						return errors.New("watch proposals unavailable")
+					}
+					retries++
+					time.Sleep(retryBackoff)
+					goto BACKOFF
```
Comment (fasmat) [inline systest/tests/common.go:391]: same as above

**Review by fasmat — CHANGES_REQUESTED** [review]

File: systest/tests/common.go (hunk 5 above)
Comment (acud) [reply systest/tests/common.go:203]: I was digging through the grpc code and the only thing I found this to be relevant to is establishing connections. Looking at their documentation, it appears that `Unavailable` is not related to establishing connections, but rather happens later in the flow. I actually doubt they would do blocking retries for application-level RPC calls, this seems out of scope for an RPC library (it is up to the application to decide whether and how to retry).

File: systest/testcontext/context.go (hunk 4 above)
Comment (acud) [reply systest/testcontext/context.go]: :+1:

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Review by fasmat — APPROVED** [review]

**Comment by acud** [comment]
bors

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

**Comment by acud** [comment]
bors try

File: systest/testcontext/context.go [hunk 9]
```diff
@@ -365,3 +375,12 @@ func New(t *testing.T, opts ...Opt) *Context {
...
 }
+
+func (c *Context) CheckFail() error {
+	select {
+	case <-failed:
+		return errors.New("test suite failed. aborting test execution")
+	default:
+	}
+	return nil
+}
```
Comment (poszu) [inline systest/testcontext/context.go:386]: Wouldn't `atomic.Bool` be more straightforward to use?

```go
failed atomic.Bool
...
t.Cleanup(func() {
	if t.Failed() {
		failed.Set(true)
	}
})
```

```suggestion
func (c *Context) CheckFail() error {
	if failed.Load() {
		return errors.New("test suite failed. aborting test execution")
	}
	return nil
}
```

**Comment by acud** [comment]
bors merge

**Pull request closed without merging**